# Intrusion Detection Evaluation Dataset (CIC-IDS2017)
By AliK604 

Intrusion Detection Systems (IDSs) and Intrusion Prevention Systems (IPSs) are the most important defense tools against the sophisticated and ever-growing network attacks. Due to the lack of reliable test and validation datasets, anomaly-based intrusion detection approaches are suffering from consistent and accurate performance evolutions.

In [1]:
# %config IPCompleter.greedy=True
import pandas as pd
import seaborn as sns
import numpy as np

import matplotlib as matplot
import matplotlib.pyplot as plt
# %matplotlib inline

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

import warnings, os 
# warnings.filterwarnings("ignore")

# from keras import Sequential
# from keras.models import Model, load_model
# from keras.layers import *
# from keras.callbacks import ModelCheckpoint
# from keras import regularizers

from sklearn.metrics import *
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, VotingClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

from sklearn.feature_selection import RFE
from sklearn.decomposition import PCA, TruncatedSVD, PCA
from sklearn.svm import LinearSVC

import xgboost, lightgbm
from mlxtend.classifier import EnsembleVoteClassifier 

/usr/local/lib/python3.6/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)


In [2]:
ls = [] 
for filename in os.listdir(r'./'):
  if '.csv' in  filename:
    print(filename)
    df = pd.read_csv(filename)
    ls.append(df)
    print(f'Shape: {df.shape}. Attack Type {df[" Label"].unique()}')

Friday-WorkingHours-Morning.pcap_ISCX.csv
Shape: (191033, 79). Attack Type ['BENIGN' 'Bot']
Monday-WorkingHours.pcap_ISCX.csv
Shape: (529918, 79). Attack Type ['BENIGN']
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Shape: (286467, 79). Attack Type ['BENIGN' 'PortScan']
Tuesday-WorkingHours.pcap_ISCX.csv
Shape: (445909, 79). Attack Type ['BENIGN' 'FTP-Patator' 'SSH-Patator']
Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Shape: (225745, 79). Attack Type ['BENIGN' 'DDoS']
Wednesday-workingHours.pcap_ISCX.csv
Shape: (692703, 79). Attack Type ['BENIGN' 'DoS slowloris' 'DoS Slowhttptest' 'DoS Hulk' 'DoS GoldenEye'
 'Heartbleed']
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Shape: (288602, 79). Attack Type ['BENIGN' 'Infiltration']
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Shape: (170366, 79). Attack Type ['BENIGN' 'Web Attack � Brute Force' 'Web Attack � XSS'
 'Web Attack � Sql Injection']


In [3]:
for df in ls:
  cols = df.columns.to_list()
  for i in range(len(cols)):
    cols[i] = cols[i].strip()
  df.columns = cols

In [4]:
df = pd.concat(ls)
df.head(3)
assert df.shape[1] == 79

Destination Port  Flow Duration  ...  Idle Min   Label
0              3268      112740690  ...  15400000  BENIGN
1               389      112740560  ...  15400000  BENIGN
2                 0      113757377  ...   5504997  BENIGN

[3 rows x 79 columns]

In [5]:
# mix of ints and floats. Label is a object (words)
# with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    # df.dtypes


In [6]:
from collections import Counter

Counter(df["Label"])

Counter({'BENIGN': 2273097,
         'Bot': 1966,
         'DDoS': 128027,
         'DoS GoldenEye': 10293,
         'DoS Hulk': 231073,
         'DoS Slowhttptest': 5499,
         'DoS slowloris': 5796,
         'FTP-Patator': 7938,
         'Heartbleed': 11,
         'Infiltration': 36,
         'PortScan': 158930,
         'SSH-Patator': 5897,
         'Web Attack � Brute Force': 1507,
         'Web Attack � Sql Injection': 21,
         'Web Attack � XSS': 652})

In [7]:
print(f'df.shape {df.shape} before sampling out most of benign data')
df = pd.concat([df[df['Label'] != 'BENIGN'], df[df['Label'] == 'BENIGN'].sample(frac=.1, random_state=42)]) # 20% of total is Malicious; Use only a sample of the BENIGN data
print(f'df.shape {df.shape} after sampling out most of benign data')

df.shape (2830743, 79) before sampling out most of benign data
df.shape (784956, 79) after sampling out most of benign data


In [8]:
le = LabelEncoder()
df['Label'] = le.fit_transform(df['Label'])
df.head(3)
le.classes_

Destination Port  Flow Duration  ...  Idle Min  Label
24072              8080       60202640  ...  10100000      1
24677              8080          57891  ...         0      1
56376              8080         134812  ...         0      1

[3 rows x 79 columns]

array(['BENIGN', 'Bot', 'DDoS', 'DoS GoldenEye', 'DoS Hulk',
       'DoS Slowhttptest', 'DoS slowloris', 'FTP-Patator', 'Heartbleed',
       'Infiltration', 'PortScan', 'SSH-Patator',
       'Web Attack � Brute Force', 'Web Attack � Sql Injection',
       'Web Attack � XSS'], dtype=object)

In [9]:
lowSTD = list(df.std().to_frame().nsmallest(20, columns=0).index)
df[lowSTD].head(3)

Bwd PSH Flags  Bwd URG Flags  ...  Down/Up Ratio  Label
24072              0              0  ...              1      1
24677              0              0  ...              1      1
56376              0              0  ...              0      1

[3 rows x 20 columns]

In [10]:
lowCORR = list(df.corr().abs().sort_values('Label')['Label'].nsmallest(20).index) # .where(lambda x: x < 0.005).dropna()
df[lowCORR].head(3)

Bwd Header Length  Fwd Header Length  ...   Bwd IAT Std  Flow Packets/s
24072                296                296  ...  4.645137e+06        0.298990
24677                 32                 32  ...  0.000000e+00       34.547684
56376                 72                 92  ...  9.154417e+04       51.924161

[3 rows x 20 columns]

In [11]:
print(f'Intersection: {set(lowSTD).intersection(set(lowCORR))}')
print(f'Union:        {len(set(lowSTD).union(set(lowCORR)))}')

Intersection: {'ECE Flag Count', 'FIN Flag Count', 'RST Flag Count', 'CWE Flag Count', 'Fwd URG Flags'}
Union:        35


In [12]:
import gc 
gc.collect()

11

In [13]:
df.dropna(inplace=True)
indices_to_keep = ~df.isin([np.nan, np.inf, -np.inf]).any(1)
df = df[indices_to_keep]


for i in df.columns:
    df = df[df[i] != "Infinity"]
    df = df[df[i] != np.nan]
    df = df[df[i] != np.inf]
    df = df[df[i] != -np.inf]
    df = df[df[i] != ",,"]
    df = df[df[i] != ", ,"]
    
print(np.any(np.isnan(df)))
print(np.any(np.isfinite(df)))

X_train, X_test, y_train, y_test = train_test_split(df.drop(['Label'],axis=1), df['Label'], test_size=.20, random_state=42)
X_train.head(2)
y_train.head(2)

False
True


Destination Port  Flow Duration  ...  Idle Max  Idle Min
51164                443         280071  ...         0         0
148374                80       85573048  ...  85400000  85400000

[2 rows x 78 columns]

51164     0
148374    4
Name: Label, dtype: int64

In [14]:
def benchmark(X_train=X_train, X_test=X_test, y_train=y_train, y_test=y_test):
  GBM = lightgbm.LGBMClassifier(objective='binary', n_estimators= 50)
  clfs = [GBM]
  for clf in clfs:
      _ = clf.fit(X_train, y_train)
      pred = clf.score(X_test, y_test)
      name = str(type(clf)).split(".")[-1][:-2]
      print("Acc: %0.5f for the %s" % (pred, name))
  return clfs

In [15]:
print('Baseline with all features')
clfs = benchmark()

Baseline with all features
Acc: 0.98591 for the LGBMClassifier


In [16]:
print('Solely with features identified as useless')

GBM = clfs[0]
remove = X_train.columns.to_numpy()[GBM.feature_importances_ == 0] 
clfs = benchmark(X_train[remove], X_test[remove])
print(remove)

Solely with features identified as useless
Acc: 0.95991 for the LGBMClassifier
['Bwd PSH Flags' 'Bwd URG Flags' 'SYN Flag Count' 'RST Flag Count'
 'CWE Flag Count' 'ECE Flag Count' 'Avg Fwd Segment Size'
 'Avg Bwd Segment Size' 'Fwd Header Length.1' 'Fwd Avg Bytes/Bulk'
 'Fwd Avg Packets/Bulk' 'Fwd Avg Bulk Rate' 'Bwd Avg Bytes/Bulk'
 'Bwd Avg Packets/Bulk' 'Bwd Avg Bulk Rate' 'Subflow Fwd Packets'
 'Subflow Fwd Bytes' 'Subflow Bwd Packets' 'Subflow Bwd Bytes']


In [17]:
print('Solely with features identified as useful')
clfs = benchmark(X_train.drop(remove,1), X_test.drop(remove,1))

Solely with features identified as useful
Acc: 0.96422 for the LGBMClassifier


In [18]:
X_train.reset_index(drop=True, inplace=True)
X_test.reset_index(drop=True, inplace=True)

print('PCA 5')
pca = PCA(5)
_ = benchmark(pca.fit_transform(X_train), pca.transform(X_test))

PCA 5
Acc: 0.92331 for the LGBMClassifier


In [19]:
print('PCA 15')
pca = PCA(15)
_ = benchmark(pca.fit_transform(X_train), pca.transform(X_test))

PCA 15
Acc: 0.95175 for the LGBMClassifier


In [20]:
print('PCA 25')
pca = PCA(25)
_ = benchmark(pca.fit_transform(X_train), pca.transform(X_test))

PCA 25
Acc: 0.96008 for the LGBMClassifier


In [21]:
tmp = np.argsort(GBM.feature_importances_)
tmp = tmp[::-1] # large to small 
top = tmp[:20] # index of the top N
rest = tmp[20:]
print(GBM.feature_importances_[top]) # check 

[1715 1540 1300 1250 1109  940  767  721  667  608  550  530  505  449
  406  403  387  362  345  338]


In [22]:
print('Top 20 features (per `LGBMClassifier`) + PCA(15) of remaining') # 2nd highest 
pca = PCA(15)

a = pd.concat([X_train.iloc[:, top], pd.DataFrame(pca.fit_transform(X_train.iloc[:, rest]))], axis=1, ignore_index=True) 
b = pd.concat([X_test.iloc[:, top] , pd.DataFrame(pca.transform    (X_test.iloc[:, rest])) ], axis=1, ignore_index=True) 

_ = benchmark(a, b)

Top 20 features (per `LGBMClassifier`) + PCA(15) of remaining
Acc: 0.98277 for the LGBMClassifier


# Conclusion 

**Baseline with all features**
* Acc: 0.98591 for the LGBMClassifier

Solely with features identified as useless
* Acc: 0.95991 for the LGBMClassifier

Solely with features identified as useful
* Acc: 0.96422 for the LGBMClassifier

PCA 5
* Acc: 0.92331 for the LGBMClassifier

PCA 15
* Acc: 0.94629 for the LGBMClassifier

PCA 25
* Acc: 0.94210 for the LGBMClassifier

**Top 20 features (per `LGBMClassifier`) + PCA(15) of remaining**
* Acc: 0.98277 for the LGBMClassifier